akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala

```diff
@@ -0,0 +1,128 @@
+      Thread.sleep(50)
+      val value2 = snapshot.asInstanceOf[MutableState].value
+      if (value1 != value2)
+        Future.failed(new IllegalStateException(s"State changed from $value1 to $value2"))
+      else {
+        state = state.updated(metadata.persistenceId, (snapshot, metadata))
+        Future.successful(())
+      }
+    }
+
+    def deleteAsync(metadata: SnapshotMetadata) = ???
+    def deleteAsync(persistenceId: String, criteria: SnapshotSelectionCriteria) = ???
+  }
+
+  def conf: Config = ConfigFactory.parseString(
+    s"""
+    akka.loglevel = INFO
+    akka.persistence.journal.leveldb.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
+    akka.persistence.snapshot-store.plugin = "slow-snapshot-store"
+
+    slow-snapshot-store.class = "${classOf[SlowInMemorySnapshotStore].getName}"
+    """)
+
+  sealed trait Command
+  final case object Increment extends Command
+  final case class GetValue(replyTo: ActorRef[Int]) extends Command
+
+  sealed trait Event
+  case object Incremented extends Event
+
+  final class MutableState(var value: Int)
+
+  def counter(
+    persistenceId: PersistenceId,
+    snapshotProbe: ActorRef[String]): EventSourcedBehavior[Command, Event, MutableState] = {
+    EventSourcedBehavior[Command, Event, MutableState](
+      persistenceId,
+      emptyState = new MutableState(0),
+      commandHandler = (state, cmd) ⇒ cmd match {
+        case Increment ⇒
+          Effect.persist(Incremented)
+
+        case GetValue(replyTo) ⇒
+          replyTo ! state.value
+          Effect.none
+      },
+      eventHandler = (state, evt) ⇒ evt match {
+        case Incremented ⇒
+          state.value += 1
+          state
+      }).onSnapshot {
+        case (meta, Success(_)) ⇒ snapshotProbe ! s"snapshot-success-${meta.sequenceNr}"
+        case (meta, Failure(_)) ⇒ snapshotProbe ! s"snapshot-failure-${meta.sequenceNr}"
+      }
+  }
+
+}
+
+class SnapshotMutableStateSpec extends ScalaTestWithActorTestKit(SnapshotMutableStateSpec.conf) with WordSpecLike {
+
+  import SnapshotMutableStateSpec._
+
+  private implicit val testSettings = TestKitSettings(system)
+
+  val pidCounter = new AtomicInteger(0)
+  private def nextPid(): PersistenceId = PersistenceId(s"c${pidCounter.incrementAndGet()})")
+
+  "A typed persistent actor with mutable state" must {
+
+    "support mutable state by stashing commands while storing snapshot" in {
+      val pid = nextPid()
+      val snapshotProbe = TestProbe[String]()
+      val snapshotState3: Behavior[Command] =
+        counter(pid, snapshotProbe.ref).snapshotWhen { (state, _, _) ⇒ state.value == 3 }
+      val c = spawn(snapshotState3)
+
+      (1 to 5).foreach { _ ⇒
+        c ! Increment
+      }
+      snapshotProbe.expectMessage(s"snapshot-success-3")
+
+      val replyProbe = TestProbe[Int]()
+      c ! GetValue(replyProbe.ref)
+      replyProbe.expectMessage(5)
+    }
+  }
+}
```

Comment:
Am I missing how this test somehow verifies ordering, shouldn't Increment cmd and evt handler tell the probe so that we can verify that 4, 5 happened after the snapshot-success?

Reply:
it did by checking for concurrent modifications but that was very vague and implicit. I have improved the test in 5c31502